File: d1/solution_two.py
INITIAL_VAL = 50
CIRCLE_SIZE = 100
INPUT_PATH = 'input.txt'


def read_input():
    with open(INPUT_PATH, 'r') as file:
        data = [line.strip() for line in file.readlines()]
        int_data = [int(line[1:]) if line[0] == 'R' else -
                    int(line[1:]) for line in data]  # now direction matters
    return int_data


def count_passing_zeros(int_list: list, initial_value: int = INITIAL_VAL, clock_size: int = CIRCLE_SIZE):
    k = initial_value
    z = int(k == 0)

    for val in int_list:
        k += val
        if k < 0:
            if val == k:
                z -= 1  # if we start from zero, we have already counted 'passing' on the prev step
            # count how many 0s crossed (neg and one-based)
            z -= k // clock_size
            k = k % clock_size  # keep k within 0 - clock_size-1
            if k == 0:            # <-- patch: count zero if landed exactly at 0
                z += 1

        elif k >= clock_size:
            # count how many 0s crossed (pos and zero-based)
            z += k // clock_size
            k = k % clock_size  # keep k within 0 - clock_size-1
        elif k == 0:
            z += 1
    return z


def dummy_count_passing_zeros(int_list: list, initial_value: int = INITIAL_VAL, clock_size: int = CIRCLE_SIZE):
    k = initial_value
    z = int(k == 0)

    for val in int_list:
        k += val
        if k == 0:
            z += 1
        while k < 0:
            if k != val:
                z += 1
            k += clock_size
            if k == 0:
                z += 1
        while k >= clock_size:
            k -= clock_size
            z += 1
    return z


def main():
    input_data = read_input()
    print(dummy_count_passing_zeros(input_data))


if __name__ == "__main__":
    main()

File: d1/solution_two.jl
const INITIAL_VAL = 50
const CIRCLE_SIZE = 100
const INPUT_PATH = "input.txt"

function read_input(path::String = INPUT_PATH)::Vector{Int}
	lines = Int[]
	open(path, "r") do file
		for line in eachline(file)
			if line[1] == 'R' # always single quotation for char
				push!(lines, parse(Int, line[2:end]))
			else
				push!(lines, -parse(Int, line[2:end]))
			end
		end
	end

	return lines
end

function count_passing_zeros(
	input::Vector{Int},
	initial_val::Int = INITIAL_VAL,
	circle_size::Int = CIRCLE_SIZE)::Int
	counter = 0
	if initial_val == 0
		counter += 1
	end

	for value in input
		initial_val += value
		if initial_val < 0
			if initial_val == value
				counter -= 1
			end
			counter -= fld(initial_val, circle_size)
			initial_val = mod(initial_val, circle_size)
		end

		if initial_val == 0
			counter += 1
		end

		if initial_val >= circle_size
			counter += fld(initial_val, circle_size)
			initial_val = mod(initial_val, circle_size)
		end
	end

	return counter
end

# Main function
function main()
	input_list = read_input()
	ans = count_passing_zeros(input_list)
	println(ans)
end

# Run main if script is executed directly
if abspath(PROGRAM_FILE) == @__FILE__
	main()
end
